src/main/java/org/broadinstitute/consent/http/db/DataAccessRequestDAO.java

```diff
@@ -124,12 +125,52 @@ AND LOWER(e.election_type) = 'dataaccess'
         AND LOWER(v.type) = 'final') final_access_vote ON
           final_access_vote.reference_id = dar.reference_id AND
           final_access_vote.dataset_id = dd.dataset_id
-      WHERE dar.submission_date > now() - interval '1 year'
-      AND final_access_vote.last_vote = TRUE
+      WHERE final_access_vote.last_vote = TRUE
       AND dar.reference_id = :darReferenceId
       AND (LOWER(dar.data->>'status') != 'archived' OR dar.data->>'status' IS NULL)
       """)
-  Set<Integer> findApprovedDatasetsByDar(@Bind("darReferenceId") String darReferenceId);
+  Set<Integer> findDatasetApprovalsByDar(@Bind("darReferenceId") String darReferenceId);
+
+  /**
+   * This query finds dataset ids on dar-dataset combinations where the most recent vote is true.
+   * This includes datasets that are a part of expired DARs, UNLIKE findApprovedDARsByDatasetId.
+   * The query accomplishes this by creating a view that is a grouping
+   * of election reference ids and LAST vote in the group of final votes for all data access
+   * elections. We need to group them due to the case of multiple elections on a dar-dataset
+   * request. Election 1 may have been denied. Election 2 may have been approved. Election 3 may
+   * have been denied again. When we partition over the election reference id, we'll get all final
+   * votes. The `LAST_VALUE` function selects the last result in the partition, which would be
+   * `FALSE` in the example above. Outside the JOIN, we filter on groupings where the final vote
+   * value is `TRUE` so the denied election in the example would be filtered out.
+   *
+   * @param darReferenceIds The DARs reference UUIDs
+   * @return Set of approved Dataset Ids for the list of DARs
+   */
+  @SqlQuery(
+      """
+      SELECT dd.dataset_id
+      FROM data_access_request dar
+      INNER JOIN dar_dataset dd ON dd.reference_id = dar.reference_id
+      INNER JOIN (
+        SELECT DISTINCT e.reference_id, e.dataset_id, LAST_VALUE(v.vote)
+        OVER(
+          PARTITION BY e.dataset_id
+            ORDER BY v.createdate
+            RANGE BETWEEN
+              UNBOUNDED PRECEDING AND
+              UNBOUNDED FOLLOWING
+        ) last_vote
+        FROM election e
+        INNER JOIN vote v ON e.election_id = v.electionid AND v.vote IS NOT NULL
+        AND LOWER(e.election_type) = 'dataaccess'
+        AND LOWER(v.type) = 'final') final_access_vote ON
+          final_access_vote.reference_id = dar.reference_id AND
+          final_access_vote.dataset_id = dd.dataset_id
+      WHERE final_access_vote.last_vote = TRUE
+        AND dar.reference_id IN (<darReferenceIds>)
```

Comment:
The main difference between findDatasetApprovalsByDar and findDatasetApprovalsByDars is that you can pass in multiple dar reference ids here, right? And it looks like this is the only line that is different.  

Is there an opportunity to share the code for this query between these two methods? 
I actually think this line would be valid even in the case that there is only a single reference id. Do we even need two queries? We could just wrap the singleton case in a list.

Reply:
> Do we even need two queries? We could just wrap the singleton case in a list.

I think this is a good idea. 

The issue we face with re-using string blocks in queries is that semgrep considers that an anti-pattern prone to sql injection problems. There might be ways around that with sql functions, but we haven't dived into that yet. Another potential solution would be to use multiple queries that are more legible.